File: app/src/main/java/com/boardgamegeek/sorter/RankSorter.java
package com.boardgamegeek.sorter;

import android.content.Context;
import android.database.Cursor;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.util.SparseArray;

import com.boardgamegeek.R;
import com.boardgamegeek.provider.BggContract.Games;

public class RankSorter extends CollectionSorter {
	@NonNull private final String defaultHeaderText;
	@NonNull private final String defaultText;
	private static final SparseArray<String> RANKS = buildRanks();

	public RankSorter(@NonNull Context context) {
		super(context);
		defaultHeaderText = context.getResources().getString(R.string.unranked);
		defaultText = context.getResources().getString(R.string.text_not_available);
	}

	@StringRes
	@Override
	protected int getDescriptionId() {
		return R.string.collection_sort_rank;
	}

	@StringRes
	@Override
	public int getTypeResource() {
		return R.string.collection_sort_type_rank;
	}

	@Override
	protected String getSortColumn() {
		return Games.GAME_RANK;
	}

	@Override
	public String getHeaderText(@NonNull Cursor cursor) {
		int rank = getInt(cursor, Games.GAME_RANK, Integer.MAX_VALUE);
		for (int i = 0; i < RANKS.size(); i++) {
			int key = RANKS.keyAt(i);
			if (rank <= key) {
				return RANKS.get(key);
			}
		}
		return defaultHeaderText;
	}

	@NonNull
	@Override
	public String getDisplayInfo(@NonNull Cursor cursor) {
		int rank = getInt(cursor, Games.GAME_RANK, Integer.MAX_VALUE);
		if (rank == Integer.MAX_VALUE) {
			return defaultText;
		}
		return String.valueOf(rank);
	}

	@NonNull
	private static SparseArray<String> buildRanks() {
		SparseArray<String> ranks = new SparseArray<>();
		ranks.put(100, "1 - 100");
		ranks.put(250, "101 - 250");
		ranks.put(500, "251 - 500");
		ranks.put(1000, "501 - 1000");
		ranks.put(2500, "1001 - 2500");
		ranks.put(5000, "2501 - 5000");
		ranks.put(10000, "5001 - 10000");
		ranks.put(Integer.MAX_VALUE, "10001+");
		return ranks;
	}
}

File: app/src/main/java/com/boardgamegeek/sorter/RankSorter.kt
package com.boardgamegeek.sorter

import android.content.Context
import android.util.SparseArray
import androidx.annotation.StringRes
import com.boardgamegeek.R
import com.boardgamegeek.model.CollectionItem
import java.text.NumberFormat

class RankSorter(context: Context) : CollectionSorter(context) {
    private val defaultHeaderText = context.resources.getString(R.string.unranked)
    private val defaultText = context.resources.getString(R.string.text_not_available)

    override val ascendingSortTypeResId: Int
        @StringRes
        get() = R.string.collection_sort_type_rank

    override val descendingSortTypeResId: Int
        @StringRes
        get() = R.string.collection_sort_type_rank_desc

    override val descriptionResId: Int
        @StringRes
        get() = R.string.collection_sort_rank

    override fun sortAscending(items: Iterable<CollectionItem>) = items.sortedBy { it.rank }

    override fun sortDescending(items: Iterable<CollectionItem>) = items.sortedByDescending { it.rank }

    override fun getHeaderText(item: CollectionItem): String {
        return (0 until ranks.size())
                .map { ranks.keyAt(it) }
                .firstOrNull { item.rank <= it }
                ?.let { ranks.get(it) }
                ?: defaultHeaderText
    }

    override fun getDisplayInfo(item: CollectionItem): String {
        return if (item.rank == CollectionItem.RANK_UNKNOWN) {
            defaultText
        } else NumberFormat.getIntegerInstance().format(item.rank)
    }

    companion object {
        private val ranks = buildRanks()

        private fun buildRanks(): SparseArray<String> {
            val rankSteps = listOf(100, 250, 500, 1000, 2500, 5000, 10000)
            val ranks = SparseArray<String>()
            for (i in rankSteps.indices) {
                ranks.put(rankSteps[i], String.format("%,d - %,d", (rankSteps.getOrElse(i - 1) { 0 }) + 1, rankSteps[i]))
            }
            ranks.put(CollectionItem.RANK_UNKNOWN - 1, String.format("%,d+", rankSteps.last() + 1))
            return ranks
        }
    }
}
